src/controllers/post/searchPosts.ts

```diff
@@ -0,0 +1,177 @@
+        invalid_type_error: "toPlace must be a valid enum of the defined places"
+      })
+      .optional(),
+
+    startTime: z
+      .coerce.date({
+        invalid_type_error: "startTime must be a Date() object"
+      })
+      .optional(),
+
+    endTime: z
+      .coerce.date({
+        invalid_type_error: "endTime must be a Date() object"
+      })
+      .optional(),
+
+    availableSeats: z
+      .number({
+        invalid_type_error: "availableSeats must be an integer"
+      })
+      .int({
+        message: "availableSeats must be an integer"
+      })
+      .nonnegative({
+        message: "availableSeats must be non-negative"
+      })
+      .optional(),
+
+    activePosts: z
+      .boolean({
+        invalid_type_error: "activePosts must be a boolean"
+      })
+      .optional(),
+
+    startAtPost: z
+      .number({
+        invalid_type_error: "startAtPost must be an integer"
+      })
+      .int({
+        message: "startAtPost must be an integer"
+      })
+      .positive({
+        message: "startAtPost must be positive"
+      })
+      .optional(),
+
+    endAtPost: z
+      .number({
+        invalid_type_error: "endAtPost must be an integer"
+      })
+      .int({
+        message: "endAtPost must be an integer"
+      })
+      .positive({
+        message: "endAtPost must be positive"
+      })
+      .optional(),
+
+    orderBy: z
+      .number({
+        invalid_type_error: "orderBy must be an integer"
+      })
+      .int({
+        message: "orderBy must be an integer"
+      })
+      .gte(-3, {
+        message: "orderBy must be an integer in the range [-3,3]"
+      })
+      .lte(3, {
+        message: "orderBy must be an integer in the range [-3,3]"
+      })
+      .optional()
+
+  })
+    .refine(data => (!(data.startTime && data.endTime) || (new Date(data.startTime) < new Date(data.endTime))),
+      "startTime must occur before endTime",
+    )
+})
+
+export const searchPostValidator = validate(dataSchema)
+
+let orderingBy: object = {
+  1: "post.createdAt",
+  2: "post.timeRangeStart",
+  3: "post.seats",
+}
+
+let orderingAlong: object = {
+  0: "ASC",
+  1: "DESC",
+  2: "DESC",
+}
+
+export const searchPosts = async (req: Request, res: Response) => {
+
+  let fromPlace: Place | null = req.body.fromPlace;
+  let toPlace: Place | null = req.body.toPlace;
+  let startTime: Date | null = req.body.startTime; // Renders trips whose timeRange is within or after startTime
+  let endTime: Date | null = req.body.endTime; // Renders trips whose timeRange is within or before endTime
+  // Use 1 or more here, to show only those posts which have available seats. leaving empty renders all posts without checking seats
+  let availableSeats: number = req.body.availableSeats || 0;
+  // true renders posts whose trips are yet to start. false renders trips which have started/finished in the past. leaving empty renders all posts regardless.
+  let activePosts: boolean | null = req.body.activePosts;
```

Comment:
Is there a reason you preferred to send all posts instead of active posts by default (so true would be the default value)? The second one sends a much smaller response right?

Reply:
There are three cases. Suppose I want to search for posts in history. Then it should be false. Or if I just want to see all my rides across time. Then its null. Or if I just want to see active posts. Then its true. Defaults to null cuz it made more sense, but again, the frontend is free to choose.

Reply:
cool